x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/PushDownVectorSimilarityFunctions.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical.local;
+
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.AttributeSet;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.FieldAttribute;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.NameId;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.FunctionEsField;
+import org.elasticsearch.xpack.esql.expression.function.vector.VectorSimilarityFunction;
+import org.elasticsearch.xpack.esql.optimizer.LocalLogicalOptimizerContext;
+import org.elasticsearch.xpack.esql.optimizer.rules.logical.OptimizerRules;
+import org.elasticsearch.xpack.esql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.local.EsqlProject;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.Attribute.rawTemporaryName;
+
+/**
+ * Replaces vector similarity functions with a field function attribute that applies
+ * the similarity function during value loading, when one side of the function is a literal.
+ * It also adds the new field function attribute to the EsRelation output, and adds a projection after it to remove it from the output.
+ */
+public class PushDownVectorSimilarityFunctions extends OptimizerRules.ParameterizedOptimizerRule<Eval, LocalLogicalOptimizerContext> {
+
+    public PushDownVectorSimilarityFunctions() {
+        super(OptimizerRules.TransformDirection.DOWN);
+    }
+
+    @Override
+    protected LogicalPlan rule(Eval eval, LocalLogicalOptimizerContext context) {
+        // TODO Extend to WHERE if necessary
+
+        AttributeSet.Builder addedAttrs = AttributeSet.builder();
+        Eval transformedEval = (Eval) eval.transformExpressionsDown(VectorSimilarityFunction.class, similarityFunction -> {
```

Comment:
This will transform both the Eval and all its upstream children. Was this intended or did you mean to use `transformExpressionsOnly`?